sv_pipeline/load_data.py

```diff
@@ -0,0 +1,392 @@
+
+        for line in tqdm(f, unit=' rows'):
+            row = line.split()
+            parsed = parse_row(row, header_indices)
+            if parsed and out_file:
+                out_file.write(line)
+
+    if out_file:
+        out_file.close()
+
+
+def subset_and_group_svs(input_dataset, sample_subset, sample_remap, ignore_missing_samples, write_subsetted_bed=False):
+    parsed_svs_by_name = {}
+    found_samples = set()
+    skipped_samples = set()
+    out_file_path = None
+    if write_subsetted_bed:
+        file_path = input_dataset.split('/')
+        file_path[-1] = 'subset_{}'.format(file_path[-1])
+        out_file_path ='/'.join(file_path)
+
+    def _parse_row(row, header_indices):
+        sample_id = get_field_val(row, SAMPLE_COL, header_indices)
+        if sample_remap and sample_id in sample_remap:
+            sample_id = sample_remap[sample_id]
+        if sample_subset is None or sample_id in sample_subset:
+            parse_sv_row(row, parsed_svs_by_name, header_indices, sample_id)
+            found_samples.add(sample_id)
+            return True
+        else:
+            skipped_samples.add(sample_id)
+            return False
+
+    load_file(input_dataset, _parse_row, out_file_path=out_file_path)
+
+    print('Found {} sample ids'.format(len(found_samples)))
+    if sample_subset:
+        if len(found_samples) != len(sample_subset):
+            missed_samples = sample_subset - found_samples
+            missing_sample_error = 'Missing the following {} samples:\n{}'.format(
+                len(missed_samples), ', '.join(sorted(missed_samples))
+            )
+            if ignore_missing_samples:
+                print(missing_sample_error)
+            else:
+                print('Samples in callset:\n{}'.format(', '.join(sorted(skipped_samples))))
+                raise Exception(missing_sample_error)
+
+    return parsed_svs_by_name
+
+
+def add_in_silico(svs_by_variant_id, file_path):
+    def _parse_row(row, header_indices):
+        variant_id = get_variant_id(row, header_indices)
+        if variant_id in svs_by_variant_id:
+            svs_by_variant_id[variant_id].update(get_parsed_column_values(row, header_indices, [IN_SILICO_COL]))
+
+    load_file(file_path, _parse_row, columns=IN_SILICO_COLS)
+
+
+def format_sv(sv):
+    sv[TRANSCRIPTS_FIELD] = [{'gene_id': gene} for gene in sv[GENES_FIELD]]
+    sv['transcriptConsequenceTerms'] = [sv[CALL_FIELD]]
+    if sv[SF_FIELD]:
+        sv['sn'] = int(sv[SC_FIELD] / sv[SF_FIELD])
+    sv['pos'] = sv[START_COL]
+    sv['xpos'] = CHROM_TO_XPOS_OFFSET[sv[CHROM_FIELD]] + sv[START_COL]
+    sv['xstart'] = sv['xpos']
+    sv['xstop'] = CHROM_TO_XPOS_OFFSET[sv[CHROM_FIELD]] + sv[END_COL]
+    sv['samples'] = []
+    for genotype in sv[GENOTYPES_FIELD]:
+        sample_id = genotype['sample_id']
+        sv['samples'].append(sample_id)
+
+        cn_key = 'samples_cn_{}'.format(genotype['cn']) if genotype['cn'] < 4 else 'samples_cn_gte_4'
```

Comment:
do we not display copy numbers greater than 4? I wouldn't know what to do with them, just curious

Reply:
We do! So we store the actual copy number on the `genotype` level object for the sample (which includes the CN and all the quality details) , and then we also store a list of sample ids where the copy number is greater than for. Thats because elasticsearch is WAY faster at answering the query "records where `cn_gte_4` contains "X" and "Y"" than "records where there is at least one nested `genotype` object with `sample_id` of either "X" or "Y"  and `CN` greater than 4". We do this for SNVs too with lists like `samples_num_alt_1`. FOR SVs we have `samples_cn_0`, `samples_cn_1`, ..., `samples_cn_3` and then everything more than that we group because for inheritance based searching (which is how you query based on CN) we consider all CNs greater than 4 to have the same inheritance pattern